rfcs/0001-messages.md

```diff
@@ -0,0 +1,301 @@
+    SameSenderAndReceiver,
+}
+```
+
+Then to create a valid `Header` in the sender side we call:
+
+```rust
+let header = Validate::validate(&Header {
+    ..
+}).expect("a valid header");
+```
+
+The receiver side will validate the header using the same method. Instead of panicking the error can be ignored to don't crash and keep waiting for other (potentially valid) messages.
+
+```rust
+if let Ok(header) = msg.header.validate() {
+    ..
+}
+```
+
+## Serialization/Deserialization
+
+Each message struct needs to serialize to bytes representation before it is sent through the wire and must deserialize to the same struct (round trip) on the receiver side. We use `serde` and macro derivations (`Serialize` and `Deserialize`) to automatically implement where possible.
+
+This will require deriving serde in several types defined in `frost.rs`. 
+Manual implementation of serialization/deserialization will be located at a new mod `src/frost/serialize.rs`.
+
+### Byte order
+
+Each byte chunk specified below is in little-endian order unless is specified otherwise.
+
+### Header
+
+The `Header` part of the message is 4 bytes total:
+
+Bytes | Field name | Data type
+------|------------|-----------
+1     | msg_type   | u8
+1     | version    | u8
+1     | sender     | u8
+1     | receiver   | u8
+
+### Primitive types
+
+`Payload`s use data types that we need to specify first. We have 2 primitive types inside the payload messages:
+
+#### `Scalar`
+
+`Scalar` is a an alias for `jubjub::Fr`. We use `Scalar::to_bytes` to get a 32-byte little-endian canonical representation. See https://github.com/zkcrypto/bls12_381/blob/main/src/scalar.rs#L252
+
+#### `AffinePoint`
+
+Much of the math in FROST is done using `jubjub::ExtendedPoint`. But for message exchange `jubjub::AffinePoint`s are a better choice, as their byte representation is smaller.
+
+Conversion from one type to the other is trivial:
+
+https://docs.rs/jubjub/0.6.0/jubjub/struct.AffinePoint.html#impl-From%3CExtendedPoint%3E
+https://docs.rs/jubjub/0.6.0/jubjub/struct.ExtendedPoint.html#impl-From%3CAffinePoint%3E
+
+We use `AffinePoint::to_bytes` to get a 32-byte little-endian canonical representation. See https://github.com/zkcrypto/jubjub/blob/main/src/lib.rs#L443
+
+### Payload
+
+Payload part of the message is variable in size and depends on message type.
+
+#### `MsgDealerBroadcast`
+
+Bytes  | Field name  | Data type
+-------|-------------|-----------
+32     | secret_key  | Scalar
+32     | commitments | AffinePoint
+32+32  | group_public| GroupPublic
+
+#### `MsgCommitments`
+
+Bytes   | Field name | Data type
+--------|------------|-----------
+32+32   | commitment | Commitment
+
+#### `MsgSigningPackage`
+
+Bytes                  | Field name     | Data type
+-----------------------|----------------|-----------
+1                      | participants   | u8
+(1+32+32)*participants | commitments    | HashMap<ParticipantID, Commitment>
+8                      | message_length | u64
```

Comment:
Shouldn't we limit the size of the messages to at most 4 GB by using u32 instead of u64? Using u64 allows the participants to set intractable lengths.

Reply:
Good question!

In general, languages and protocols are moving away from hard-coded 4GB data limits. For example, Rust uses `usize` (native pointer size) for memory sizes, and `u64` for network or file sizes.

Some threshold signature use-cases will involve "messages" that are larger than 4GB. For example, open-source software distribution would benefit from threshold signatures. And a full Debian distribution is 12.5 GB.

But for other protocols, such as Zcash, 1 GB would be an intractable length. So we can't really use integer sizes to limit message sizes - they're too coarse.

Here's what I'd recommend:
1. use `u64` in the generic protocol
2. add an implementation-specific message size limit in the `MsgSigningPackage::validate` method
3. add a "not included" note that each protocol will have its own message size limit

For Zcash, the message size limit would be 2 MB, because Zcash network messages, blocks, and transactions are limited to 2 MB.

But for modern Bitcoin, it might be 32 MB, because that's their network message size limit.